### PR TITLE
001 view job roles

### PR DESCRIPTION
**Pull request by niall-mccloy**

### CODE REVIEW

**Review by Meganc3379 — COMMENTED** [review]
Looks good

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+name: Run all tests with Maven
+
+on:
+  push:
+    branches: [ "main" ]
```
Comment (mlynch147) [inline .github/workflows/test.yml]: change this to all branches - the tests should run everytime you push to your feature branch

File: src/main/java/org/example/controllers/JobRoleController.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.sql.SQLException;
+
+@Api("Kainos Job Application - Job Roles API")
+@Path("/api/")
```
Comment (mlynch147) [inline src/main/java/org/example/controllers/JobRoleController.java]: Question - have you tested this with swagger?  I notice this class level path ends in a '/' and the method level path starts with a '/'.  If all is ok via swagger, then no issue

File: src/test/java/com/kainos/ea/controller/JobRoleControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+@ExtendWith(MockitoExtension.class)
+public class JobRoleControllerTest {
+
+    JobRoleService jobRoleService = Mockito.mock(JobRoleService.class);
+
+    JobRoleController jobRoleController = new JobRoleController(jobRoleService);
+
+    @Test
+    public void getJobRoles_shouldReturnResponseCode200WithJobRoleResponses() throws SQLException {
```
Comment (mlynch147) [inline src/test/java/com/kainos/ea/controller/JobRoleControllerTest.java]: picky i know, but from a naming POV it should be something like 

getJobRoles_shouldReturnResponseCode200WithJobRoleResponses_whenNoErrorsThrown

modify all tests so that this naming convention is observed

File: src/test/java/com/kainos/ea/controller/JobRoleControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        Response actualResponse = Response.ok().entity(jobRoleService.getAllJobRoles()).build();
+
+        // check if status codes match
+        Assert.assertEquals(expectedResponse.getStatus(), actualResponse.getStatus());
+        // check if the entity/body of expected and actual responses match
+        // i.e. check if lists of job role responses match
+        Assert.assertEquals(expectedResponse.getEntity(), actualResponse.getEntity());
+    }
+
+    // testing if the SQLException is handled correctly and an internal server error is returned.
```
Comment (mlynch147) [inline src/test/java/com/kainos/ea/controller/JobRoleControllerTest.java]: no need for comment

File: src/test/java/com/kainos/ea/dao/JobRoleDaoTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/*
```
Comment (mlynch147) [inline src/test/java/com/kainos/ea/dao/JobRoleDaoTest.java]: This whole class is commented out - and their is only a single test

File: src/test/java/com/kainos/ea/mapper/JobRoleMapperTest.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        Date closingDate = new Date(millis);
+        JobRole testJobRole = new JobRole("SE", "Derry",
+                "Eng", "Band 4", closingDate);
+        List<JobRole> jobRoles = new ArrayList<>();
+        jobRoles.add(testJobRole);
+
+        List<JobRoleResponse> jobRoleResponses = JobRoleMapper.mapJobRolesListToJobRoleResponseList(jobRoles);
+
+        Assert.assertEquals(1, jobRoleResponses.size());
+        // saving the object at index 0 in the jobRoleResponses list to a variable
```
Comment (mlynch147) [inline src/test/java/com/kainos/ea/mapper/JobRoleMapperTest.java]: no need for comment

File: src/test/java/com/kainos/ea/mapper/JobRoleMapperTest.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        JobRoleResponse response = jobRoleResponses.get(0);
+        Assert.assertEquals("SE", response.getRoleName());
+        Assert.assertEquals("Derry", response.getLocation());
+        Assert.assertEquals("Eng", response.getCapabilityName());
+        Assert.assertEquals("Band 4", response.getBandName());
+        Assert.assertEquals(closingDate, response.getClosingDate());
+
+    }
+
+    // testing edge case handling
```
Comment (mlynch147) [inline src/test/java/com/kainos/ea/mapper/JobRoleMapperTest.java]: remove comment

**Review by Meganc3379 — APPROVED** [review]
Looks good Niall!

File: src/main/java/org/example/daos/JobRoleDao.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class JobRoleDao {
+
+    public List<JobRole> getAllJobRoles()
```
Comment (mlynch147) [inline src/main/java/org/example/daos/JobRoleDao.java]: rename to getOpenJobRoles()

File: src/main/java/org/example/services/JobRoleService.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+public class JobRoleService {
+
+    JobRoleDao jobRoleDao;
+    DatabaseConnector databaseConnector;
+
+    public JobRoleService(final JobRoleDao jobRoleDao) {
+        this.jobRoleDao = jobRoleDao;
+    }
+
+    public List<JobRoleResponse> getAllJobRoles() throws SQLException {
```
Comment (mlynch147) [inline src/main/java/org/example/services/JobRoleService.java]: rename to getOpenJobRoles

File: src/test/java/com/kainos/ea/integration/JobRoleIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        List<JobRole> response = client
+                .target("http://localhost:8080/api/job-roles")
+                .request()
+                .get(List.class);
+
+        Assertions.assertFalse(response.isEmpty());
+    }
+
+    @Test
+    void getJobRoles_shouldReturnResponseCode500_whenSqlExceptionIsCaught()
```
Comment (mlynch147) [inline src/test/java/com/kainos/ea/integration/JobRoleIntegrationTest.java]: remove this test....

Comment (niall-mccloy) [reply src/test/java/com/kainos/ea/integration/JobRoleIntegrationTest.java]: Test removed

File: src/main/java/org/example/services/JobRoleService.java (hunk 9 above)
Comment (niall-mccloy) [reply src/main/java/org/example/services/JobRoleService.java]: Renamed method

File: src/main/java/org/example/daos/JobRoleDao.java (hunk 8 above)
Comment (niall-mccloy) [reply src/main/java/org/example/daos/JobRoleDao.java]: renamed method

**Review by mlynch147 — APPROVED** [review]
